Add support for a different hitobject sheet for lns

7ef38c opened the pull request:
Closes #4445 

If note-holdobject-sheet@9x1.png exists it is now used, else fall back to the hitobject sheet if enabled.

(ln heads are inverted in example)
![image](https://github.com/user-attachments/assets/19279697-5dc7-4ed5-85c2-7216d4f1f9ca)

7ef38c committed:
Add support for a dedicated snap sheet for ln

also detects if a hitobject sheet exists but no hold object sheet exists (although detection is kinda sketch so i'd like to get someone to check that before i push note to self)

7ef38c committed:
Formatting

Ebe learns how to read the codestyle guidelines colorized 2025

WilliamQiufeng commented on Quaver.Shared/Skinning/SkinKeys.cs:
```diff
@@ -742,9 +742,13 @@ private void LoadLaneSpecificElements()
+                    var holdobjects = LoadSpritesheet(SkinKeysFolder.HitObjects, "note-holdobject-sheet", false, snapCount, 1);
+
+                    if (holdobjects.Count == 1)
+                        holdobjects=objects;
```
```suggestion
                        holdobjects = objects;
```
Should be a space in between

7ef38c committed:
Update Quaver.Shared/Skinning/SkinKeys.cs

Co-authored-by: WilliamQiufeng <williamqiufeng@outlook.com>

WilliamQiufeng commented on Quaver.Shared/Skinning/SkinKeys.cs:
```diff
@@ -742,9 +742,13 @@ private void LoadLaneSpecificElements()
                     var objects = LoadSpritesheet(SkinKeysFolder.HitObjects, "note-hitobject-sheet", false, snapCount, 1);
+                    var holdobjects = LoadSpritesheet(SkinKeysFolder.HitObjects, "note-holdobject-sheet", false, snapCount, 1);
+
+                    if (holdobjects.Count == 1)
```
This prevents a 1x1 sprite sheet to be loaded. Not a big deal though since we can just override holdhitobject instead for 1x1

7ef38c committed:
Better holdobject sheet detection

unsure if rewriting to support any YxZ combination is worth it since we already expect a 9x1 sheet anyways but this should work

Warp9000 commented on Quaver.Shared/Skinning/SkinKeys.cs:
```diff
@@ -742,9 +742,13 @@ private void LoadLaneSpecificElements()
                     const int snapCount = 9;
 
                     var objects = LoadSpritesheet(SkinKeysFolder.HitObjects, "note-hitobject-sheet", false, snapCount, 1);
+                    var holdobjects = LoadSpritesheet(SkinKeysFolder.HitObjects, "note-holdobject-sheet", false, snapCount, 1);
+
+                    if (!File.Exists($"{Store.Dir}/{ModeHelper.ToShortHand(Mode).ToLower()}/{SkinKeysFolder.HitObjects}/note-holdobject-sheet@9x1.png"))
+                        holdobjects = objects;
```
sheets arent limited to 9x1. see https://github.com/Quaver/Quaver/blob/f57cec2197d3fbaaa5529860cdcb32ef321dfb3c/Quaver.Shared/Skinning/SkinStore.cs#L104

also and if else here would make it clearer whats happening

7ef38c committed:
Update SkinKeys.cs

every time i rewrite the if check to be more flexible it gets longer and i dont think thats a good thing

Warp9000 commented on Quaver.Shared/Skinning/SkinKeys.cs:
```diff
@@ -741,18 +742,21 @@ private void LoadLaneSpecificElements()
 
                     for (var j = 0; j < snapCount - NoteHitObjects[i].Count; j++)
-                        NoteHitObjects[i].Add(NoteHitObjects[i].Last());
+                            NoteHitObjects[i].Add(NoteHitObjects[i].Last());
```
?

Warp9000 commented on Quaver.Shared/Skinning/SkinKeys.cs:
```diff
@@ -741,18 +742,21 @@ private void LoadLaneSpecificElements()
 
-                    if (!File.Exists($"{Store.Dir}/{ModeHelper.ToShortHand(Mode).ToLower()}/{SkinKeysFolder.HitObjects}/note-holdobject-sheet@9x1.png"))
-                        holdobjects = objects;
+                    if (Directory.GetFiles($@"{Store.Dir}/{ModeHelper.ToShortHand(Mode).ToLower()}/{SkinKeysFolder.HitObjects}").Any(file => new Regex(SkinStore.SpritesheetRegex("note-holdobject-sheet")).IsMatch(Path.GetFileName(file))))
```
> every time i rewrite the if check to be more flexible it gets longer and i dont think thats a good thing

yeah i can see that.
something like [this](https://github.com/Quaver/Quaver/blob/f57cec2197d3fbaaa5529860cdcb32ef321dfb3c/Quaver.Shared/Skinning/SkinStore.cs#L417-L426) would be more readable.

WilliamQiufeng replied:
You can make some variables like
```cs
var dir = $@"{Store.Dir}/{ModeHelper.ToShortHand(Mode).ToLower()}/{SkinKeysFolder.HitObjects}";
var files = Directory.GetFiles(dir);
var spritesheetRegex = new Regex(SkinStore.SpritesheetRegex("note-holdobject-sheet"));
if (files.Any(file => spritesheetRegex.IsMatch(Path.GetFileName(file)))) {...}
```
(do not worry about slowing down the program by making new variables)

Making `new Regex(SkinStore.SpritesheetRegex("note-holdobject-sheet"))` its own variable outside the if statement would be better since the value is not changing inside the lambda of the `Any()`, and you have been constructing the same thing per `file` (it is also possible the the compiler knows about this already xd)

7ef38c committed:
Update SkinKeys.cs

Applied William's suggestions + fixed the indent i somehow caused

Co-Authored-By: WilliamQiufeng <williamqiufeng@outlook.com>

Warp9000 commented on Quaver.Shared/Skinning/SkinKeys.cs:
```diff
@@ -741,10 +742,21 @@ private void LoadLaneSpecificElements()
-                    NoteHitObjects.Add(objects);
-                    NoteHoldHitObjects.Add(objects);
+                    var dir = $@"{Store.Dir}/{ModeHelper.ToShortHand(Mode).ToLower()}/{SkinKeysFolder.HitObjects}";
+                    var files = Directory.GetFiles(dir);
```
`System.IO.DirectoryNotFoundException: Could not find a part of the path '/home/warp/code/Quaver/Quaver/bin/Debug/net6.0/Skins/Default Skin/4k/HitObjects'.`

7ef38c replied:
gg

7ef38c committed:
Fix the default arrow skin breaking

This should only happen with the default arrow skin as it doesnt have a folder in store.dir but this should make it behave normally?

7ef38c committed:
Update SkinKeys.cs

oh yeah that works

7ef38c committed:
Update SkinKeys.cs

why does it reindent that im sobbing

7ef38c committed:
pain

Co-Authored-By: Warp <59202433+Warp9000@users.noreply.github.com>

Warp9000 reviewed (approved):
so much work for such a small change... xd
